### PR TITLE
[WIP] Add append functionality to `faas-cli new` command

### PR DESCRIPTION
**Pull request by ericstoekl**
Appends function config YAML data to existing YAML file, specified by:
`faas-cli new <FuncName> --lang <Lang> -f <YAML>`

Signed-off-by: Eric Stoekl <ems5311@gmail.com>

This change allows for merging the YAML config data for a new function into an existing YAML file. It also adds tests (7 test cases) for the `new` function.

## Description

Instead of appending text directly to the file, we open the file with `stack.ParseYAMLFile`, append data to the returned `stack.Services` struct, then write the data back to the file with `stack.WriteYAMLData` (a new function for the `stack` package).

## Motivation and Context

- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
#128 

## How Has This Been Tested?

Added `new_function_test.go` which runs several test cases against the function, testing its original and new functionality.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: commands/new_function_test.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+		os.Remove("template")
+	}()
+
+	// Change directory to testdata
+	if err := os.Chdir("testdata/new_function"); err != nil {
+		t.Fatalf("Error on cd to testdata dir: %v", err)
+	}
+
+	// Symlink template directory at root to current directory to avoid re-downloading templates
+	os.Symlink("../../../template", "template")
```
Comment (alexellis) [inline commands/new_function_test.go]: I think this could be asking for trouble. Can we do a `cp` or just generate a minimum set of files and have that as part of the tests?

File: stack/schema.go [hunk 2]
```diff
@@ -7,13 +7,13 @@ package stack
...
 	GatewayURL string `yaml:"gateway"`
-	Network    string `yaml:"network"`
+	Network    string `yaml:"network,omitempty"`
 }
 
 // Function as deployed or built on FaaS
 type Function struct {
 	// Name of deployed function
-	Name     string `yaml:"-"`
+	Name     string `yaml:"-,omitempty"`
```
Comment (alexellis) [inline stack/schema.go]: Does this re-serialize correctly? I had some issues getting this to map back in the correct way.

File: stack/stack.go [hunk 3]
```diff
@@ -103,3 +103,16 @@ func fetchYAML(address *url.URL) ([]byte, error) {
 
 	return resBytes, err
 }
+
+func WriteYAMLData(services *Services, fileName string) error {
```
Comment (alexellis) [inline stack/stack.go]: Have we got any tests on the serialization on its own?

File: stack/schema.go (hunk 2 above)
Comment (ericstoekl) [reply stack/schema.go]: That's strange, what kind of issues did you experience? The `omitempty` is there so that if the `Name` property is left out of the `Function` struct instantiation, it will not be Marshalled to YAML. Without `omitempty`, the YAML file will end up with a `name: ""` property.

File: stack/stack.go (hunk 3 above)
Comment (ericstoekl) [reply stack/stack.go]: No but I will add some.

File: commands/new_function_test.go (hunk 1 above)
Comment (ericstoekl) [reply commands/new_function_test.go]: Sure, I'll copy the folder over instead of using `os.Symlink`.

**Comment by alexellis** [comment]
What if we wrote a minimal example in the folder structure instead? - we are only testing that the YAML gets appended - not strictly caring about the templated function itself.

**Comment by ericstoekl** [comment]
Can you clarify what you mean by "minimal example in the folder structure"? Should the test be testing only parseYAML/WriteYAML, and not doing full `faasCmd.Exec()` calls to run the whole `new` function?

**Comment by alexellis** [comment]
We're copy in a whole set of templates. Could we not just write a template folder with the minimum text to make the process work?

**Comment by ericstoekl** [comment]
Gotcha, thanks for clarifying. I'll use a minimal templates directory.

**Comment by ericstoekl** [comment]
Added tests for the `WriteYAMLData` function.

Leaving this un-squashed for now for easier reviewing (IMO), let me know if you have any other questions/issues.

File: stack/stack.go (hunk 3 above)
Comment (itscaro) [reply stack/stack.go]: @ericstoekl can you rename it to WriteYAMLDataForServices, in #87 I made some changes, ParseYAMLData is for general parsing, and ParseYAMLDataForServices is used for services

Comment (ericstoekl) [reply stack/stack.go]: Thanks for the heads up @itscaro. If #87 gets merged before this one, I'll make that update.

**Comment by developius** [comment]
What's the status on this? @ericstoekl @alexellis

**Comment by ericstoekl** [comment]
Should be good to go in my opinion, but could use some code review + testing by someone who isn't me :)

**Comment by developius** [comment]
Looks good & works fine well - tested using the `samples.yml` file. My only comments would be that it will remove reformat the yaml file (including removing comments & reordering definitions). Not a big deal but would be nice for this not to happen (think `git diff` etc).

**Comment by ericstoekl** [comment]
@developius Thanks for testing. @alexellis , like Finnian said the ordering of the yaml file will be potentially re-arranged when you append to it. This is because I use `yaml.Marshal(services)` in the `WriteYAMLData()` function, (https://github.com/ericstoekl/faas-cli/blob/0bfb0c8aa282978ebce326d277b8e65759b9eb11/stack/stack.go#L113)

This will take the `services` object, which has a schema here (https://github.com/ericstoekl/faas-cli/blob/0bfb0c8aa282978ebce326d277b8e65759b9eb11/stack/schema.go#L46). This enforces that the `provider:` section in the YAML comes before the `functions:` section. The ordering of the `Function` struct is also enforced in the YAML.

The result is that if the user has an ordering that deviates from the order that the schema has, that ordering will be re-arranged. Also, if the user has whitespace in their YAML file, it will all be removed.

Let me know if this is acceptable. If not, we can implement the append feature by making sure that the `function:` section is last, then write string data of the new function object to the end of the YAML file.

**Comment by alexellis** [comment]
This is tricky.. I know that marshalling the file is the technically correct thing to do, but appending text may be a better user experience as @developius has a good point on the comments and whitespace.

I'd like to get an opinion from @nicholasjackson and @stefanprodan

**Comment by nicholasjackson** [comment]
So I am kinda on the fence, I am pro fixing formatting and whitespace as GoFmt does and the other benefit is that using YAML parser will give early feedback that the user has borked their config.  Loosing comments however is a big one, personally I would never add comments to the file so it would be a minor problem for me however it may cause an issue to the user.  This could be avoided with a message in the help, "Appends new function to existing config, this operation will reformat your existing yaml including the removal of comments.

The other option could be as simple as read the file with the YAML parser to check correctness, use the yaml parser to create the new template, merge the two by string concatenation.  That way I think everyone wins.

**Comment by developius** [comment]
I'm all for @nicholasjackson's solution - read the file and parse to check integrity, add the new function via string concatenation and then parse again to check.

**Comment by alexellis** [comment]
@nicholasjackson the resulting ordering is also "borked" - you can end up with the provider section at the bottom for instance.

**Comment by alexellis** [comment]
@ericstoekl we have consensus on appending rather than re-writing the whole file. Would you like to open a new PR for the work?

**Comment by ericstoekl** [comment]
@alexellis shouldn't be too much of a problem to do the work in this PR, as the tests won't change too much. Is that acceptable?

**Comment by ericstoekl** [comment]
My plan for the next PR to solve this issue (would appreciate feedback if this approach is correct or not @alexellis ):

1) Parse existing YAML, make sure it's OK
2) Create `Function` struct with data for the new function user wants to create
3) Read existing YAML file line-by-line, until end of `functions:` section (until another section is found or until EOF)
4) Insert struct created in step 2 into YAML file at this point

Does that sound reasonable? I'm working on it right now, will have a new PR soon.

**Comment by developius** [comment]
Sounds good @ericstoekl but I'd add a 5th step to re-validate the yaml before inserting, just to be safe

File: commands/new_function.go [hunk 4]
```diff
@@ -25,6 +26,7 @@ func init() {
 	newFunctionCmd.Flags().StringVar(&lang, "lang", "", "Language or template to use")
 	newFunctionCmd.Flags().StringVar(&gateway, "gateway", defaultGateway,
 		"Gateway URL to store in YAML stack file")
+	newFunctionCmd.Flags().StringVar(&image, "image", "", "Name for docker image")
```
Comment (alexellis) [inline commands/new_function.go]: While an interesting change, this shouldn't be part of the PR.

Comment (alexellis) [reply commands/new_function.go]: @ericstoekl please see this feedback.

Comment (ericstoekl) [reply commands/new_function.go]: I've got a new PR on the way for this feature, which addresses this. I'll close that when it's ready.

**Comment by ericstoekl** [comment]
Re-done with new PR - #213
